<a href="https://colab.research.google.com/github/HKang42/DS-Unit-2-Applied-Modeling/blob/master/module1-define-ml-problems/Harrison_Kang_LS_DS_231_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 3, Module 1*

---


# Define ML problems

You will use your portfolio project dataset for all assignments this sprint.

## Assignment

Complete these tasks for your project, and document your decisions.

- [ ] Choose your target. Which column in your tabular dataset will you predict?
- [ ] Is your problem regression or classification?
- [ ] How is your target distributed?
    - Classification: How many classes? Are the classes imbalanced?
    - Regression: Is the target right-skewed? If so, you may want to log transform the target.
- [ ] Choose your evaluation metric(s).
    - Classification: Is your majority class frequency >= 50% and < 70% ? If so, you can just use accuracy if you want. Outside that range, accuracy could be misleading. What evaluation metric will you choose, in addition to or instead of accuracy?
    - Regression: Will you use mean absolute error, root mean squared error, R^2, or other regression metrics?
- [ ] Choose which observations you will use to train, validate, and test your model.
    - Are some observations outliers? Will you exclude them?
    - Will you do a random split or a time-based split?
- [ ] Begin to clean and explore your data.
- [ ] Begin to choose which features, if any, to exclude. Would some features "leak" future information?

If you haven't found a dataset yet, do that today. [Review requirements for your portfolio project](https://lambdaschool.github.io/ds/unit2) and choose your dataset.

Some students worry, ***what if my model isn't “good”?*** Then, [produce a detailed tribute to your wrongness. That is science!](https://twitter.com/nathanwpyle/status/1176860147223867393)

In [1]:
from google.colab import files

uploaded = files.upload()

In [0]:
import pandas as pd

songs = pd.read_csv('kaggle_songs.txt', names=['song_id', 'song_index'], sep = ' ')
users = pd.read_csv('kaggle_users.txt', names=['user_id'])
data = pd.read_csv('kaggle_visible_evaluation_triplets.txt', sep='\t', names=['user_id', 'song_id','play_count'])

In [0]:
url = "http://labrosa.ee.columbia.edu/millionsong/sites/default/files/AdditionalFiles/unique_tracks.txt"

song_map = pd.read_csv(url, sep='<SEP>', engine='python', header = None, names=['track_id', 'song_id', 'artist', 'song'])

In [11]:
# User listening histories.
# Each row is the listening history of a user, a song, and that song's playcount

data.head(10) 

user_id             song_id  play_count
0  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d  SOBONKR12A58A7A7E0         1.0
1  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d  SOEGIYH12A6D4FC0E3         1.0
2  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d  SOFLJQZ12A6D4FADA6         1.0
3  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d  SOHTKMO12AB01843B0         1.0
4  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d  SODQZCY12A6D4F9D11         1.0
5  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d  SOXLOQG12AF72A2D55         1.0
6  d7083f5e1d50c264277d624340edaaf3dc16095b  SOUVUHC12A67020E3B         1.0
7  d7083f5e1d50c264277d624340edaaf3dc16095b  SOUQERE12A58A75633         1.0
8  d7083f5e1d50c264277d624340edaaf3dc16095b  SOIPJAX12A8C141A2D         1.0
9  d7083f5e1d50c264277d624340edaaf3dc16095b  SOEFCDJ12AB0185FA0         2.0

In [5]:
# list of song_ids and song index
# song_index is just the index +1. It can probably be dropped.
songs.head()

song_id  song_index
0  SOAAADD12AB018A9DD           1
1  SOAAADE12A6D4F80CC           2
2  SOAAADF12A8C13DF62           3
3  SOAAADZ12A8C1334FB           4
4  SOAAAFI12A6D4F9C66           5

In [6]:
# list of user identifiers
# used to sort predictions in submitted file
users.head()

user_id
0  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d
1  d7083f5e1d50c264277d624340edaaf3dc16095b
2  d68dc6fc25248234590d7668a11e3335534ae4b4
3  9be82340a8b5ef32357fe5af957ccd54736ece95
4  841b2394ae3a9febbd6b06497b4a8ee8eb24b7f8

In [7]:
# maps track ID and song ID to song and artist
# track ID isn't needed
song_map.head()

track_id             song_id            artist               song
0  TRMMMYQ128F932D901  SOQMMHC12AB0180CB8  Faster Pussy cat       Silent Night
1  TRMMMKD128F425225D  SOVFVAK12A8C1350D9  Karkkiautomaatti        Tanssi vaan
2  TRMMMRX128F93187D9  SOGTUKN12AB017F4F1    Hudson Mohawke  No One Could Ever
3  TRMMMCH128F425532C  SOBNYVR12A8C13558C       Yerba Brava      Si Vos Querés
4  TRMMMWA128F426B589  SOHSBXH12A8C13B0DF        Der Mystic   Tangle Of Aspens

In [8]:
 data.shape, songs.shape, users.shape, song_map.shape

((773813, 3), (386213, 2), (110000, 1), (1000000, 4))

In [14]:
# how many unique users are in our data?
data['user_id'].nunique()

58699

In [9]:
# Let's see what songs these users are listening to
data['song_id'].value_counts()

SOFRQTD12A81C233C0    2733
SOAUWYT12A81C206F1    2433
SOBONKR12A58A7A7E0    2213
SOSXLTC12AF72A7F54    2008
SOAXGDH12A8C13F8A1    2007
                      ... 
SOZHABY12A8C139465       1
SOOXCIM12A58A7920E       1
SOADCOJ12A8C1353AA       1
SOOXQPB12AB0189145       1
SOMYRCU12A6D4FAC97       1
Name: song_id, Length: 125239, dtype: int64

In [12]:
# check some of the songs
song_map[song_map['song_id'] == 'SOBONKR12A58A7A7E0']

track_id             song_id         artist            song
448841  TRAEHHJ12903CF492F  SOBONKR12A58A7A7E0  Dwight Yoakam  You're The One

#Choose your target. Which column in your tabular dataset will you predict?
- Is your problem regression or classification?
- How is your target distributed?
- Classification: How many classes? Are the classes imbalanced?
- Regression: Is the target right-skewed? If so, you may want to log transform the target.

I will predict which song a user will listen to.

This is a classification problem with 1,000,000 possible options. However, I suspect that a few songs will dominate playlists

# Choose your evaluation metric(s).
- Classification: Is your majority class frequency >= 50% and < 70% ? If so, you can just use accuracy if you want. Outside that range, accuracy could be misleading. What evaluation metric will you choose, in addition to or instead of accuracy?
- Regression: Will you use mean absolute error, root mean squared error, R^2, or other regression metrics?


I think accuracy may be the best or most practical evaluation metric. Calculating precision or recall with so many songs might not be very useful. However, these metrics could be useful if I were to change the problem to a binary "will this person listen to a specific song" problem. 

 # Choose which observations you will use to train, validate, and test your model.
- Are some observations outliers? Will you exclude them?
- Will you do a random split or a time-based split?

The 'data' dataframe which contains 773,813 observations detailing the number of times a specific user has listened to a specific song. There is no ordering to the observations. So the data set will be randomly split. May need to consider not allowing individual users to be split across sets.